File: start_server.py
#!/usr/bin/env python3
"""
NexGame Casino - Local Development Server
Simple HTTP server for running the casino game locally
"""

import http.server
import socketserver
import webbrowser
import os
import sys
from pathlib import Path

# Configuration
PORT = 8000
HOST = 'localhost'

class CustomHTTPRequestHandler(http.server.SimpleHTTPRequestHandler):
    """Custom handler to serve files with proper MIME types"""
    
    def end_headers(self):
        # Add CORS headers for development
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        super().end_headers()
    
    def guess_type(self, path):
        """Override to set correct MIME types"""
        result = super().guess_type(path)
        if isinstance(result, tuple):
            mimetype, encoding = result
        else:
            mimetype, encoding = result, None
        
        # Ensure proper MIME types for game assets
        if path.endswith('.js'):
            return 'application/javascript'
        elif path.endswith('.css'):
            return 'text/css'
        elif path.endswith('.html'):
            return 'text/html'
        elif path.endswith('.png'):
            return 'image/png'
        elif path.endswith('.jpg') or path.endswith('.jpeg'):
            return 'image/jpeg'
        elif path.endswith('.gif'):
            return 'image/gif'
        elif path.endswith('.svg'):
            return 'image/svg+xml'
        
        return mimetype

def find_free_port(start_port=8000, max_port=8100):
    """Find a free port starting from start_port"""
    for port in range(start_port, max_port):
        try:
            with socketserver.TCPServer((HOST, port), CustomHTTPRequestHandler) as server:
                return port
        except OSError:
            continue
    return None

def main():
    """Main function to start the server"""
    print("🎰 NexGame Casino - Development Server")
    print("=" * 40)
    
    # Check if index.html exists
    if not os.path.exists('index.html'):
        print("❌ Error: index.html not found!")
        print("   Make sure you're running this script from the game directory.")
        sys.exit(1)
    
    # Find a free port
    port = find_free_port(PORT)
    if port is None:
        print(f"❌ Error: No free ports available in range {PORT}-8100")
        sys.exit(1)
    
    # Start the server
    try:
        with socketserver.TCPServer((HOST, port), CustomHTTPRequestHandler) as httpd:
            print(f"🚀 Server started successfully!")
            print(f"📍 URL: http://{HOST}:{port}")
            print(f"📁 Serving files from: {os.getcwd()}")
            print()
            print("🎮 Game Controls:")
            print("   Space/Enter: Spin")
            print("   A: Auto Spin")
            print("   +/-: Change Bet")
            print("   Escape: Close Win Modal")
            print()
            print("⚙️  Admin Panel:")
            print("   Click the gear icon (⚙️) to access admin controls")
            print()
            print("🛑 Press Ctrl+C to stop the server")
            print("=" * 40)
            
            # Open browser automatically
            try:
                webbrowser.open(f'http://{HOST}:{port}')
                print("🌐 Opening game in browser...")
            except Exception as e:
                print(f"⚠️  Could not open browser automatically: {e}")
                print(f"   Please open http://{HOST}:{port} manually")
            
            print()
            
            # Start serving
            httpd.serve_forever()
            
    except KeyboardInterrupt:
        print("\n🛑 Server stopped by user")
    except OSError as e:
        if e.errno == 48:  # Address already in use
            print(f"❌ Error: Port {port} is already in use")
            print("   Try running the script again or use a different port")
        else:
            print(f"❌ Error starting server: {e}")
        sys.exit(1)
    except Exception as e:
        print(f"❌ Unexpected error: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: start_server.js
#!/usr/bin/env node
/**
 * NexGame Casino - Local Development Server
 * Node.js HTTP server for running the casino game locally
 */

const http = require('http');
const fs = require('fs');
const path = require('path');
const { exec } = require('child_process');

// Configuration
const PORT = 8000;
const HOST = 'localhost';

// MIME types
const mimeTypes = {
    '.html': 'text/html',
    '.css': 'text/css',
    '.js': 'application/javascript',
    '.json': 'application/json',
    '.png': 'image/png',
    '.jpg': 'image/jpeg',
    '.jpeg': 'image/jpeg',
    '.gif': 'image/gif',
    '.svg': 'image/svg+xml',
    '.ico': 'image/x-icon',
    '.woff': 'font/woff',
    '.woff2': 'font/woff2',
    '.ttf': 'font/ttf',
    '.eot': 'application/vnd.ms-fontobject'
};

// Find free port
function findFreePort(startPort = PORT, maxPort = 8100) {
    return new Promise((resolve, reject) => {
        const server = http.createServer();
        
        server.listen(startPort, HOST, () => {
            const port = server.address().port;
            server.close(() => resolve(port));
        });
        
        server.on('error', (err) => {
            if (err.code === 'EADDRINUSE') {
                if (startPort < maxPort) {
                    findFreePort(startPort + 1, maxPort).then(resolve).catch(reject);
                } else {
                    reject(new Error(`No free ports available in range ${PORT}-${maxPort}`));
                }
            } else {
                reject(err);
            }
        });
    });
}

// Get MIME type
function getMimeType(filePath) {
    const ext = path.extname(filePath).toLowerCase();
    return mimeTypes[ext] || 'application/octet-stream';
}

// Serve file
function serveFile(filePath, res) {
    const mimeType = getMimeType(filePath);
    
    fs.readFile(filePath, (err, data) => {
        if (err) {
            res.writeHead(404, { 'Content-Type': 'text/plain' });
            res.end('File not found');
            return;
        }
        
        res.writeHead(200, { 
            'Content-Type': mimeType,
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
            'Access-Control-Allow-Headers': 'Content-Type'
        });
        res.end(data);
    });
}

// Handle request
function handleRequest(req, res) {
    let filePath = '.' + req.url;
    
    // Default to index.html for root
    if (filePath === './') {
        filePath = './index.html';
    }
    
    // Check if file exists
    fs.access(filePath, fs.constants.F_OK, (err) => {
        if (err) {
            // Try index.html for directory requests
            if (filePath.endsWith('/')) {
                filePath += 'index.html';
                fs.access(filePath, fs.constants.F_OK, (err2) => {
                    if (err2) {
                        res.writeHead(404, { 'Content-Type': 'text/plain' });
                        res.end('File not found');
                    } else {
                        serveFile(filePath, res);
                    }
                });
            } else {
                res.writeHead(404, { 'Content-Type': 'text/plain' });
                res.end('File not found');
            }
        } else {
            serveFile(filePath, res);
        }
    });
}

// Open browser
function openBrowser(url) {
    const start = process.platform === 'darwin' ? 'open' :
                  process.platform === 'win32' ? 'start' : 'xdg-open';
    
    exec(`${start} ${url}`, (error) => {
        if (error) {
            console.log(`⚠️  Could not open browser automatically: ${error.message}`);
            console.log(`   Please open ${url} manually`);
        }
    });
}

// Main function
async function main() {
    console.log('🎰 NexGame Casino - Development Server');
    console.log('=' .repeat(40));
    
    // Check if index.html exists
    if (!fs.existsSync('index.html')) {
        console.log('❌ Error: index.html not found!');
        console.log('   Make sure you\'re running this script from the game directory.');
        process.exit(1);
    }
    
    try {
        // Find free port
        const port = await findFreePort();
        
        // Create server
        const server = http.createServer(handleRequest);
        
        // Start server
        server.listen(port, HOST, () => {
            console.log(`🚀 Server started successfully!`);
            console.log(`📍 URL: http://${HOST}:${port}`);
            console.log(`📁 Serving files from: ${process.cwd()}`);
            console.log();
            console.log('🎮 Game Controls:');
            console.log('   Space/Enter: Spin');
            console.log('   A: Auto Spin');
            console.log('   +/-: Change Bet');
            console.log('   Escape: Close Win Modal');
            console.log();
            console.log('⚙️  Admin Panel:');
            console.log('   Click the gear icon (⚙️) to access admin controls');
            console.log();
            console.log('🛑 Press Ctrl+C to stop the server');
            console.log('=' .repeat(40));
            
            // Open browser
            openBrowser(`http://${HOST}:${port}`);
            console.log();
        });
        
        // Handle server errors
        server.on('error', (err) => {
            if (err.code === 'EADDRINUSE') {
                console.log(`❌ Error: Port ${port} is already in use`);
                console.log('   Try running the script again or use a different port');
            } else {
                console.log(`❌ Error starting server: ${err.message}`);
            }
            process.exit(1);
        });
        
        // Handle graceful shutdown
        process.on('SIGINT', () => {
            console.log('\n🛑 Server stopped by user');
            server.close(() => {
                process.exit(0);
            });
        });
        
    } catch (error) {
        console.log(`❌ Error: ${error.message}`);
        process.exit(1);
    }
}

// Run if called directly
if (require.main === module) {
    main();
}

module.exports = { main, findFreePort, handleRequest };
